src/main.rs

```diff
@@ -1088,20 +1117,38 @@ fn read_events(path: &Path, rdesc: &ReportDescriptor) -> Result<()> {
     let mut start_time: Option<Instant> = None;
     let mut last_timestamp: Option<Instant> = None;
     let mut data = [0; 1024];
+
+    let mut builder = libbpf_rs::RingBufferBuilder::new();
+    builder.add(map_ringbuf, event_handler).unwrap();
+    let ringbuf = builder.build().unwrap();
+
     loop {
-        let mut pollfds = [PollFd::new(f.as_fd(), PollFlags::POLLIN)];
+        let ringbuf_fd = unsafe {
+            std::os::fd::BorrowedFd::borrow_raw(ringbuf.epoll_fd() as std::os::fd::RawFd)
+        };
+        let mut pollfds = [
+            PollFd::new(f.as_fd(), PollFlags::POLLIN),
+            PollFd::new(ringbuf_fd, PollFlags::POLLIN),
+        ];
+
         if poll(&mut pollfds, timeout)? > 0 {
             match f.read(&mut data) {
                 Ok(_nbytes) => {
                     last_timestamp = print_current_time(last_timestamp);
                     if start_time.is_none() {
                         start_time = last_timestamp;
                     }
-                    parse_input_report(&data, rdesc, &start_time.unwrap())?;
+                    parse_input_report(&data, rdesc, &start_time.unwrap(), &ringbuf)?;
                 }
                 Err(e) => {
                     if e.kind() != std::io::ErrorKind::WouldBlock {
                         bail!(e);
+                    } else {
+                        last_timestamp = print_current_time(last_timestamp);
+                        if start_time.is_none() {
+                            start_time = last_timestamp;
+                        }
+                        let _ = ringbuf.consume();
```

Comment:
honestly, this is a big ugly. We're expecting an error from the fd read and then fall back to the bpf fd. The correct approach for `poll()` would be to check which one of the `pollfds` has `PollFlags::revents()` set to `PollFlags::POLLIN`. So something like:

```rust
if poll(&mut pollfds, timeout)? > 0 {
   if pollfds[0].revents().map_or(false, |f| f.intersects(PollFlags::POLLIN)) {
       match f.read() {
       }
   }
   // same for pollfds[1]
}
```

(I typed this here so it probably won't compile like this but you get the gist)

Reply:
done in #34, please feel free to cherry-pick and squash